components/fxa-client/src/lib.rs

```diff
@@ -141,6 +141,40 @@ impl FirefoxAccount {
             .append_pair("showSuccessMessage", "true");
         Ok(url)
     }
+
+    /// Get the "manage account" page URL.
+    /// It is typically used in the application's account status UI,
+    /// to link the user out to a webpage where they can manage
+    /// all the details of their account.
+    ///    
+    /// * `entrypoint` - Application-provided string identifying the UI touchpoint
+    ///                  through which the page was accessed, for metrics purposes.
+    pub fn get_manage_account_url(&mut self, entrypoint: &str) -> Result<Url> {
+        let mut url = self.state.config.content_url_path("settings")?;
+        url.query_pairs_mut().append_pair("entrypoint", entrypoint);
+        self.add_account_identifiers_to_url(url)
+    }
+
+    /// Get the "manage devices" page URL.
+    /// It is typically used in the application's account status UI,
+    /// to link the user out to a webpage where they can manage
+    /// the devices connected to their account.
+    ///
+    /// * `entrypoint` - Application-provided string identifying the UI touchpoint
+    ///                  through which the page was accessed, for metrics purposes.
+    pub fn get_manage_devices_url(&mut self, entrypoint: &str) -> Result<Url> {
+        let mut url = self.state.config.content_url_path("settings/clients")?;
+        url.query_pairs_mut().append_pair("entrypoint", entrypoint);
+        self.add_account_identifiers_to_url(url)
+    }
+
+    fn add_account_identifiers_to_url(&mut self, mut url: Url) -> Result<Url> {
+        let profile = self.get_profile(false)?;
```

Comment:
These have to take a `&mut self` in order to be able to get the `uid` and `email`, because it might need to fetch that over the network and update the local cache.  I guess technically this means they should also try persist state on completion?

Something about having to do `get_profile` here doesn't sit right with me, but I'm not sure what the best alternative is.  I supposed I could add some logic for just looking directly in `self.profile_cache`, but I wonder if it's worth using this as an opportunity to refactor `uid` into top-level account data rather than being tucked away in the profile blob.

Reply:
>  I guess technically this means they should also try persist state on completion?

Nope as the profile cache is in-memory only.

> Something about having to do get_profile here doesn't sit right with me, but I'm not sure what the best alternative is. I supposed I could add some logic for just looking directly in self.profile_cache, but I wonder if it's worth using this as an opportunity to refactor uid into top-level account data rather than being tucked away in the profile blob.

Agree, maybe we could do this once we try to pull the Kotlin bits over to Rust later in the year?

Reply:
It's probably fair to move the cache behind a RefCell to make conceptually non-mutable things not require &mut.

Reply:
Agree, I just didn't know much about interior mutability when I first worked on that crate

Reply:
> It's probably fair to move the cache behind a RefCell to make conceptually
> non-mutable things not require &mut.

Do I also need to use a `Mutex` or similar to ensure that concurrent access to the `RefCell` doesn't panic, or do we have magic to take care of that elsewhere?

Reply:
The FFI prevents concurrent access.

(Also, RefCell isn't safe for concurrent access anyway, although Rust should tell you if you ever try to use something non-threadsafe in an environment that needs that. Opting out of this requires `unsafe`, either in Rust or in our FFI)

FWIW it's probably fine for us to fix this in a follow up, I was just commenting that maybe we should do that.

Reply:
I had a shot at using `RefCell` but it had a lot of followon effects, e.g. `get_access_token` also takes `&mut self` and it wasn't obvious whether we should change that.